### PR TITLE
feat(parser,evaluator): implement string-type flag semantics (#234)

### LINKED ISSUE #234: Implement string-type flag semantics (/c /C /w /W /B /b /t /T)
## Summary

magic(5) defines flag suffixes on `string` types that alter comparison semantics:

- `/c` -- case-insensitive, uppercase letters in the magic value match either case in the file
- `/C` -- case-insensitive, lowercase letters in the magic value match either case in the file
- `/w` -- whitespace-optional: whitespace bytes in the magic value are optional in the file
- `/W` -- compact-whitespace: multiple whitespace in the magic value matches any amount of whitespace in the file
- `/t` -- force-text: file is treated as text for MIME purposes
- `/T` -- force-binary: file is treated as binary
- `/b` / `/B` -- blank-handling variants (specifics vary by libmagic version)

## Current state (after PR #233)

The parser **accepts and parses-and-drops** all flag letters in `parse_type_and_operator` for `string`-family types (e.g., `string/w =EXFAT` in `/usr/share/file/magic/filesystems` line 265 loads cleanly). However, the evaluator does **not** alter `read_string_exact` / `read_string` / `apply_equal` behavior based on the flags. The match still uses byte-exact comparison.

This means rules that rely on `/c` (case-insensitive) or `/w`/`/W` (whitespace flexibility) will fail to match files that GNU `file` would correctly identify, even though they parse without error.

## Real-world need

`/usr/share/file/magic/filesystems` line 265: `3 string/w =EXFAT` -- ExFAT detection with whitespace flexibility. Currently parses but match correctness depends on the buffer not having weird whitespace, which is the case for ExFAT headers (no whitespace at all), so this specific rule happens to work by coincidence. Other `/c`/`/C` rules in the broader corpus (Python script detection, magic-file headers) will silently fail.

## Implementation outline

1. **AST** -- Extend `TypeKind::String { ... }` with a `flags: StringFlags` field (or a new `string_flags` field on `MagicRule`). New `StringFlags` struct with bool fields for each flag.
2. **Parser** -- The current parse-and-drop in `parser/grammar/mod.rs` (after the `pstring` / `regex` / `search` suffix branches) should populate the new field instead of discarding the consumed letters. Tag tracking PR #233.
3. **Evaluator** -- New `read_string_with_flags` (or extend `read_string_exact`) that:
   - Walks both pattern and buffer in parallel when `/w`/`/W` is set, allowing variable whitespace runs in the buffer to match a single whitespace in the pattern.
   - Uses `eq_ignore_ascii_case` (or a more careful Unicode case-fold) for `/c`/`/C` letter matches. magic(5) `/c` and `/C` differ in which side controls -- read libmagic's `softmagic.c` for the exact contract.
4. **Format substitution** -- `/t`/`/T` are MIME hints, not match-altering. May only need to be honored when emitting MIME output (out of scope for this issue if MIME isn't yet emitted).
5. **Codegen** -- `serialize_type_kind` for the new field.
6. **Tests** -- Each flag combination needs at least one positive and one negative match test.

## Acceptance criteria

- [ ] Each flag in `/c /C /w /W /B /b` alters comparison semantics correctly per magic(5)
- [ ] Flag combinations work (e.g., `string/cw`)
- [ ] `/t` / `/T` flags are recorded but no comparison change (or wired into MIME output if that lands first)
- [ ] Round-trip codegen
- [ ] At least one fixture-based conformance test per flag against GNU `file` reference output

## Refs

- Epic #54 (Type System Expansion)
- PR #233 (parse-and-drop landed; semantic implementation deferred)
- magic(5) man page, "String types" section

### PR DESCRIPTION
**Pull request by unclesp1d3r**
## Summary

Implements `string`-type flag semantics (#234), the v0.3.0 prerequisite tracked for the EPUB / archive-rule coverage in #51.

Each magic(5) flag letter now alters comparison behavior per libmagic `src/softmagic.c`:

- **`/c` / `/C`** — ASCII case-insensitive with the libmagic asymmetric contract (pattern char controls fold direction). See GOTCHAS S6.5 for the mixed-case behavior.
- **`/w` / `/W`** — whitespace-optional / whitespace-required-compact via parallel-walk. The consumed-bytes count drives the relative-offset anchor for child rules (load-bearing for chained matches).
- **`/T`** — `STRING_TRIM`: leading/trailing ASCII whitespace stripped from the pattern.
- **`/f`** — `STRING_FULL_WORD`: post-match word-boundary check.
- **`/t` / `/b`** — text/binary hints, parsed and stored; output-side wiring is gated on `!:mime` evaluation (deferred to #51).

## Issue body corrections landed alongside the implementation

Research against libmagic source surfaced three errors in the original #234 description:

1. `/T` is `STRING_TRIM` (whitespace trim), not "force-binary" as the issue claimed.
2. `/B` is **not** a string flag — it is the `pstring` 1-byte length-width letter per `src/file.h`. The grammar now rejects `string/B` at parse time. (A regression guard test exercises the `parse_text_magic_file` path.)
3. `/f` (`STRING_FULL_WORD`) existed in libmagic but was missing from the issue's enumeration.

AGENTS.md "Current Limitations" updated; GOTCHAS.md S6.5 (asymmetric `/c`/`/C`) and S6.6 (`/B` exclusion) added.

## Design

- AST shape `TypeKind::String { max_length, flags: StringFlags }` mirrors the existing `RegexFlags` precedent.
- Default flags (`StringFlags::is_empty()`) take the byte-exact fast path through `apply_equal` — no perf regression for plain `string` rules.
- Non-default flags route through the pattern-bearing-type contract (GOTCHAS S2.4) via `read_pattern_match` and the new `compare_string_with_flags`. This mirrors the regex/search dispatch and gives ordering-operator rejection for free.
- `/T` trim happens at evaluation time (in `read_pattern_match`) rather than at AST construction — simpler, no `Value::String` mutation. Documented deviation from the plan.

## Test coverage

| Layer | Where | Coverage |
|---|---|---|
| Algorithm | `src/evaluator/types/string.rs::compare_string_with_flags` | 12 unit tests: per-flag, combinations, asymmetric `/c`/`/C`, boundary cases, the load-bearing consumed-bytes contract |
| Engine dispatch | `src/evaluator/engine/tests/mod.rs` | 6 tests: routing decision, NotEqual polarity, ordering-operator rejection, default-flag fast path, `/W` anchor advance for relative-offset children |
| Grammar | `src/parser/grammar/tests/mod.rs` | Per-letter mapping (table-driven), combinations, `/B` rejection |
| Integration / conformance | `tests/string_flags_integration.rs` | 9 tests against real-world rule shapes (ExFAT, Python shebang, FTCOMP, trim, full-word, combined) |
| Codegen round-trip | `src/build_helpers.rs` | Default + non-default flags emit + recompile |
| Proptest | `tests/property_tests.rs` | `arb_type_kind` now generates non-default `StringFlags` for serde round-trip |

## Test plan

- [x] `cargo test --all-targets` — 1611 tests + 194 doc tests pass
- [x] `cargo clippy --all-targets -- -D warnings` — clean
- [x] `cargo fmt --check` — clean
- [x] Built-in rules regenerate successfully (`StringFlags` imported in generated `builtin_rules.rs`)
- [ ] Manual cross-check against GNU `file` on representative rules — deferred to reviewer / follow-up

## Plan

`docs/plans/2026-05-26-001-feat-string-type-flag-semantics-plan.md` (gitignored; lives locally)

## Downstream

This PR unblocks the `string/b FTCOMP` archive rule in #51's coverage requirement and lays the foundation for #235 (`search`-type flag semantics — `/s` and `/b` share most of the comparison contract).

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -231,7 +231,7 @@ See **Development Phases** below for the planned roadmap of features not yet imp
...
-- `string` type modifier flags are not supported: `/B` (compact whitespace), `/b` (compact blanks), `/c`/`/C` (case-insensitive), `/t`/`/T` (force text/binary), `/w`/`/W` (whitespace optional). Only `pstring` has suffix parsing today.
+- `string` type modifier flag *parser* accepts `/c /C /w /W /b /t /T /f` but the evaluator still uses byte-exact comparison; semantics are not yet honored (tracked in issue #234). `/B` is **not** a `string` flag in libmagic — it is exclusively the `pstring` 1-byte length-width letter — so `string/B` is rejected at parse time. Note: an earlier draft of this file incorrectly listed `/T` as "force-binary"; per libmagic `src/file.h`, `/T` is `STRING_TRIM` (leading/trailing whitespace trim) and `/b` is `STRING_BINTEST` (force-binary). Bare `/f` is `STRING_FULL_WORD` (post-match word-boundary check).
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply AGENTS.md]: > `Current Limitations` states that `string` flag semantics are not yet honored... Please update this section to reflect that `/c /C /w /W /T /f` now affect comparison/anchor advancement, while `/t` and `/b` are currently stored as MIME hints (no comparison effect), and `string/B` is rejected.

Addressed: the AGENTS.md "Current Limitations" entry was rewritten in PR #288 commit eafdd4a (line 234 of the merged state) to describe the implemented behavior. The new wording covers `/c /C /w /W /T /f` semantics, the `/t` / `/b` MIME-hint scope, and the `string/B` rejection.

File: src/parser/ast.rs [hunk 2]
```diff
@@ -749,6 +761,160 @@ impl RegexFlags {
...
+    /// comparison. Lowercase pattern chars are compared literally.
+    pub ignore_uppercase: bool,
+    /// `/t` -- `STRING_TEXTTEST`. Hint that this rule applies to text
+    /// files. Captured for MIME-output integration; does not currently
+    /// alter comparison.
+    pub text_test: bool,
+    /// `/T` -- `STRING_TRIM`. Trim leading and trailing ASCII whitespace
+    /// from the pattern before comparison. The trim happens at AST
+    /// construction time; the comparison function receives the trimmed
+    /// pattern.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/parser/ast.rs]: > `StringFlags::trim` docs say trimming happens at AST construction time, but the implementation trims at evaluation time

Addressed: doc updated in PR #288 commit eafdd4a (the `pub trim: bool` field doc in `src/parser/ast.rs` now says "The trim is applied at evaluation time (in `read_pattern_match`)").

File: src/evaluator/types/string.rs [hunk 3]
```diff
@@ -125,6 +125,124 @@ pub fn read_string_exact(
...
+/// and GOTCHAS S6.5 for the canonical contract. This function implements the
+/// libmagic `src/softmagic.c` algorithm verbatim (in spirit) using
+/// `u8::to_ascii_lowercase` / `to_ascii_uppercase` for the fold so the
+/// comparison stays ASCII-only and locale-independent.
+///
+/// # Arguments
+///
+/// * `pattern` -- The magic rule's pattern bytes, already trimmed when
+///   `/T` was set (the AST-construction step is the trim point).
+/// * `buffer` -- The file slice (full buffer; this function indexes from
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/types/string.rs]: > The rustdoc for `compare_string_with_flags` says trim happens at AST construction time

Addressed: rustdoc rewritten in PR #288 commit ef8efbf — the function now says "Trim is applied by the caller, not here. `read_pattern_match` (in the parent module) trims the pattern before invoking this function when `flags.trim` is set."

File: src/evaluator/types/mod.rs [hunk 4]
```diff
@@ -330,6 +377,60 @@ pub(crate) fn read_pattern_match(
...
 
+/// Trim leading and trailing ASCII whitespace from a byte slice.
+///
+/// Used for the `/T` (`STRING_TRIM`) flag on `string` rules. ASCII-only
+/// trim matches libmagic's `isspace`-based contract; full Unicode
+/// whitespace handling is out of scope.
+/// Anchor-advance count for a flagged `string` rule.
+///
+/// Flagged string rules go through the pattern-bearing-type contract (see
+/// `read_pattern_match`), so their anchor advance is whatever
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/types/mod.rs]: > The doc comment immediately above `trim_ascii_whitespace` accidentally includes the text "Anchor-advance count for a flagged `string` rule."

Addressed: the misplaced rustdoc was moved in PR #288 commit ef8efbf. `trim_ascii_whitespace` now carries its own correct doc, and the "Anchor-advance count..." paragraph is attached to `flagged_string_bytes_consumed` where it belongs.

File: tests/string_flags_integration.rs [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+#[test]
+fn string_t_trims_pattern_whitespace() {
+    // Pattern has leading and trailing spaces; trim should match against
+    // a buffer that contains only the inner content. This proves the
+    // trim happened at evaluation time before the comparison.
+    let r = rule(
+        0,
+        "  hello  ",
+        StringFlags::default().with_trim(true),
+        "hello",
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply tests/string_flags_integration.rs:142]: > Test name `string_t_trims_pattern_whitespace` is easy to misread as the `/t` flag, but it's validating `/T` (trim)

Addressed in follow-up PR #293: renamed to `string_capital_t_trims_pattern_whitespace`. The other two `/T` tests got the same prefix.

File: src/evaluator/types/mod.rs [hunk 6]
```diff
@@ -330,6 +377,60 @@ pub(crate) fn read_pattern_match(
...
+            );
+            return 0;
+        }
+    };
+    let effective: &[u8] = if flags.trim {
+        trim_ascii_whitespace(pattern_bytes)
+    } else {
+        pattern_bytes
+    };
+    string::compare_string_with_flags(effective, buffer, offset, flags).unwrap_or(0)
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/types/mod.rs:440]: > `flagged_string_bytes_consumed` currently returns the raw `compare_string_with_flags` consumption and does not apply the `TypeKind::String` "consume a trailing NUL terminator if present" behavior

Addressed in follow-up PR #293: `flagged_string_bytes_consumed` now peeks `scan_buffer[offset+consumed]` and includes the NUL when present, mirroring the unflagged path. Pinned by a new regression test `relative_after_flagged_string_parent_includes_nul_terminator`.

File: src/evaluator/types/mod.rs [hunk 7]
```diff
@@ -321,6 +331,43 @@ pub(crate) fn read_pattern_match(
...
+        // and rejects other operators on the rule.
+        //
+        // Pattern can be `Value::String` (the common case) or `Value::Bytes`
+        // (parser-emitted for backslash-escape literals like `\177ELF`).
+        // The trim flag (`/T`) is honored here at evaluation time so the AST
+        // construction stays unchanged.
+        TypeKind::String { flags, .. } if !flags.is_empty() => {
+            let pattern_bytes: &[u8] = match pattern {
+                Some(Value::String(s)) => s.as_bytes(),
+                Some(Value::Bytes(b)) => b.as_slice(),
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/types/mod.rs:347]: > Flagged-string matching in `read_pattern_match` ignores `max_length` on `TypeKind::String`

Addressed in follow-up PR #293: both `read_pattern_match` and `flagged_string_bytes_consumed` now clamp the scan buffer to `offset..offset+n` when `max_length: Some(n)`. Pinned by a new test `flagged_string_respects_max_length_cap`.

File: tests/string_flags_integration.rs [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+// ---------- Regression guard: /T with all-whitespace pattern ----------
+
+/// `string/T "   "` would silently match every file if we let the empty
+/// post-trim pattern through to `compare_string_with_flags` (which
+/// returns `Some(0)` for an empty pattern -- the same hazard documented
+/// in GOTCHAS S2.5 for regex). The fix in `read_pattern_match` rejects
+/// the empty-after-trim case explicitly with `TypeReadError::UnsupportedType`,
+/// which the engine surfaces via `evaluate_rules` as no match (rather
+/// than a panic or universal-match).
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply tests/string_flags_integration.rs:245]: > The comment claims `read_pattern_match` rejects the empty-after-trim pattern with `TypeReadError::UnsupportedType`, but the current implementation logs a warning and returns `Ok(None)`

Addressed in follow-up PR #293: comment updated to "logs a `warn!` and returns `Ok(None)` (no match) so the malformed rule surfaces in logs without aborting evaluation of subsequent rules in the file."

File: src/evaluator/types/mod.rs [hunk 9]
```diff
@@ -321,6 +331,61 @@ pub(crate) fn read_pattern_match(
...
+        // Flagged `string` rules go through the pattern-bearing-type
+        // contract (GOTCHAS S2.4): on hit return `Some(Value::String(
+        // matched_bytes))`, on miss return `None`. The engine dispatcher
+        // (`evaluate_pattern_rule`) translates Some/None into Equal/NotEqual
+        // and rejects other operators on the rule.
+        //
+        // Pattern can be `Value::String` (the common case) or `Value::Bytes`
+        // (parser-emitted for backslash-escape literals like `\177ELF`).
+        // The trim flag (`/T`) is honored here at evaluation time so the AST
+        // construction stays unchanged.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/types/mod.rs:343]: > The doc comment for flagged `string` rules says a hit returns `Some(Value::String(matched_bytes))`, but the code actually returns `Some(Value::Bytes(...))` to preserve byte-exact semantics.

Addressed in follow-up PR #293: doc updated to say `Value::Bytes(matched_bytes)` and explain why (libmagic byte-exact semantics; `from_utf8_lossy` would corrupt non-UTF-8). Cross-references GOTCHAS S2.3 (cross-type String/Bytes equality policy).

File: src/evaluator/strength.rs [hunk 10]
```diff
@@ -71,11 +71,22 @@ pub fn calculate_default_strength(rule: &MagicRule) -> i32 {
...
-            // Base string strength
+        // Strings are most specific (they match exact byte sequences).
+        // Flagged strings get a per-flag penalty because the flags broaden
+        // what the rule matches: `string/c FOO` matches both `FOO` and
+        // `foo`, so it should sort BELOW `string FOO` (which only matches
+        // `FOO` exactly) under stop_at_first_match. This mirrors libmagic's
+        // `apprentice.c::apprentice_magic_strength` which subtracts 1 per
+        // `STRING_IGNORE_*` flag bit. We extend the penalty to whitespace
+        // and full-word flags by the same logic: each flag that makes the
+        // pattern fuzzier costs one point.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/strength.rs:82]: > This comment says the strength penalty is extended to "whitespace and full-word flags", but `string_flag_specificity_penalty()` does not penalize `/f`

Addressed in follow-up PR #293: comment rewritten to spell out penalized (`/c`, `/C`, `/w`, `/W`) vs non-penalized (`/T`, `/t`, `/b`, `/f`) with rationale ("`/f` is NOT penalized because it tightens the match (requires a word boundary) rather than broadening it").

File: src/evaluator/strength.rs [hunk 11]
```diff
@@ -199,6 +210,37 @@ pub fn calculate_default_strength(rule: &MagicRule) -> i32 {
     strength.clamp(MIN_STRENGTH, MAX_STRENGTH)
 }
 
+/// Count how many `string`-flag bits make a rule's pattern fuzzier.
+///
+/// libmagic `apprentice.c::apprentice_magic_strength` subtracts 1 from the
+/// computed strength for each `STRING_IGNORE_*` bit set. We extend the same
+/// reasoning to whitespace and full-word flags: each flag that broadens
+/// what the rule matches reduces its specificity by one point.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/strength.rs:218]: > The helper doc says the penalty logic extends to "whitespace and full-word flags", but `/f` is explicitly *not* penalized

Addressed in follow-up PR #293: helper rustdoc rewritten with explicit penalized/non-penalized lists matching the implementation.

File: docs/src/ast-structures.md [hunk 12]
```diff
@@ -439,6 +443,71 @@ let case_start = TypeKind::Regex {
...
+// Combined flags (case-insensitive + whitespace-optional)
+let combined = TypeKind::String {
+    max_length: Some(128),
+    flags: StringFlags::default()
+        .with_ignore_lowercase(true)
+        .with_compact_optional_whitespace(true),
+};
+```
+
+`StringFlags::default()` constructs a value with every flag set to `false`. The `is_empty(&self) -> bool` method returns `true` for such a value, and the engine uses it as the fast-path predicate: rules with default flags take the byte-exact comparison path, while any non-default flag routes the rule through `compare_string_with_flags`.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply docs/src/ast-structures.md:507]: > This doc references an `is_empty(&self) -> bool` method, but `StringFlags::is_empty` is currently defined as `pub const fn is_empty(self) -> bool` (takes `self` by value).

Addressed in follow-up PR #293: doc updated to `is_empty(self) -> bool` with a note that `StringFlags: Copy` so taking `self` is cheap.

File: docs/MAGIC_FORMAT.md [hunk 13]
```diff
@@ -214,26 +214,70 @@ Examples:
 0       pstring/HJ =JPEG        JPEG image (2-byte BE, self-inclusive length)
 ```
 
-If `max_length` is specified in the magic file (not shown in the basic syntax), it caps the length value to prevent reading excessive data.
+If `max_length` is specified in the magic file (not shown in the basic syntax), it caps the length value to prevent reading excessive data. This guards against attacker-controlled length-prefix saturation attacks where malicious files specify extreme length values.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply docs/MAGIC_FORMAT.md:217]: > This paragraph implies a `max_length` can be specified in the text magic file for `pstring`, but the current grammar only supports `/B/H/h/L/l` and optional `/J`

Addressed in follow-up PR #293: paragraph rewritten to clarify the pstring magic-file syntax accepts only `/B/H/h/L/l/J` and that the AST `max_length` field is reserved for programmatic construction. Rules loaded from `.magic` text always have `max_length: None`.

File: docs/src/magic-format.md [hunk 14]
```diff
@@ -283,22 +283,52 @@ The optional max_length parameter caps the length value:
 0       pstring   x         \b, name: %s
 ```
 
-### String Flags (Not Yet Implemented)
+### String Flags
 
-> **Note:** String flags are documented for libmagic compatibility reference but are not yet implemented in libmagic-rs.
+String flags are now implemented as of PR #234, providing libmagic-compatible string comparison semantics.
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply docs/src/magic-format.md:288]: > This sentence says string flags are implemented "as of PR #234", but other updated docs in this PR refer to the implementation landing in PR #288 (and #234 is the tracking issue).

Addressed in follow-up PR #293: changed to "implemented (issue #234, landed in PR #288)" for consistency with AGENTS.md and GOTCHAS.md.

File: src/evaluator/strength.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/strength.rs:82]: > Line 80-82 says full-word flags are penalized, but the implementation does not penalize `/f`.

Addressed in follow-up PR #293 — same fix as the Copilot `/f` thread above. The comment now spells out penalized (`/c`, `/C`, `/w`, `/W`) vs non-penalized (`/T`, `/t`, `/b`, `/f`) explicitly.

File: src/evaluator/strength.rs [hunk 15]
```diff
@@ -199,6 +210,37 @@ pub fn calculate_default_strength(rule: &MagicRule) -> i32 {
...
+        penalty += 1;
+    }
+    if flags.compact_whitespace {
+        penalty += 1;
+    }
+    if flags.compact_optional_whitespace {
+        penalty += 1;
+    }
+    penalty
+}
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/strength.rs:242]: > This helper changes rule ordering semantics, but there are no explicit cases here asserting penalized (`/c`, `/C`, `/w`, `/W`) vs non-penalized (`/T`, `/b`, `/t`, `/f`) behavior. Add a small table-driven test to lock this contract.

Addressed in follow-up PR #293: added `string_flag_specificity_penalty_per_flag_table` — a table-driven unit test in `src/evaluator/strength.rs::tests` with 12 rows covering single-flag, multi-flag-stacking, and mixed penalized + non-penalized cases.

File: src/evaluator/engine/tests/mod.rs [hunk 16]
```diff
@@ -2748,6 +2755,149 @@ fn test_search_parent_relative_child_at_positive_offset() {
...
+        EvaluationContext::new(EvaluationConfig::default().with_stop_at_first_match(false));
+    // File: "a    b!" -> parent consumes 6 bytes, child reads byte at
+    // offset 6 which is '!'. The contract from U3 test
+    // `test_compare_string_with_flags_consumed_bytes_drives_anchor` is
+    // what makes this work end-to-end.
+    let matches = evaluate_rules(&[parent], b"a    b!", &mut context).unwrap();
+    assert_eq!(matches.len(), 2, "parent + child must both match");
+    assert_eq!(matches[1].message, "exclaim");
+}
+
```
[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/engine/tests/mod.rs:2900]: > Please move this section (helper + related tests) into a focused submodule (for example, `src/evaluator/engine/tests/string_flags_dispatch_tests.rs`)

Deferred to a separate refactor PR: the file-size concern is valid (`src/evaluator/engine/tests/mod.rs` is ~2900 lines), but the bloat predates PR #288 — the test file was already oversized before the string-flag section was added. Splitting it into focused submodules is a mechanical refactor that should be its own PR rather than expanding the scope of an in-flight feature change. Tracking the broader concern is implicit in the existing AGENTS.md file-size guidance; we'll address it when next touching this module substantially.

[the remark this answers is not in the document]
Comment (unclesp1d3r) [reply src/evaluator/engine/tests/mod.rs:2900]: Update: actually addressed in follow-up PR #295. The helper + 6 tests moved to a new `src/evaluator/engine/tests/string_flags_dispatch_tests.rs` submodule; parent `tests/mod.rs` dropped from 2920 to 2786 lines.
